sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -634,7 +634,12 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
     if (!p.equals(lastRegex)) {
       // regex value changed
       lastRegex = p.asInstanceOf[UTF8String].clone()
-      pattern = Pattern.compile(lastRegex.toString)
+      try {
+        pattern = Pattern.compile(lastRegex.toString)
+      } catch {
+        case e: PatternSyntaxException =>
+          throw QueryExecutionErrors.invalidPatternError(prettyName, e.getPattern)
```

Comment:
it would be useful to have another version of this method passing `e` as the cause to SparkRuntimeException because it contains useful diagnostics for the user:
```
java.util.regex.PatternSyntaxException: Unclosed counted closure near index 8
[a\d]{0, 2}
```

Reply:
What about other instances of `Pattern.compile` in this class  such as `regexp_extract`?

also `rlike` does not hide PatternSyntaxException but does not use  QueryExecutionErrors.invalidPatternError
```
scala> sql("select rlike('', '[a\\\\d]{0, 2}')").collect()
java.util.regex.PatternSyntaxException: Unclosed counted closure near index 8
[a\d]{0, 2}
```